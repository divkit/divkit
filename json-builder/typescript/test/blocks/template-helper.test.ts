import {
    DivContainer,
    DivGallery,
    DivImage,
    DivImageBackground,
    DivLinearGradient,
    DivSolidBackground,
    DivText,
    fixed,
    reference,
    template,
    templateHelper,
    expression,
} from '../../src';

describe('templateHelper', () => {
    it('should create TemplateBlock with parameters', () => {
        const templates = {
            template1: new DivContainer({
                paddings: reference('vPaddings'),
                items: [
                    new DivText({
                        text: 'text',
                        max_lines: reference('vMaxlines'),
                    }),
                    new DivText({
                        text: reference('vText1'),
                    }),
                    new DivText({
                        text: reference('vText2'),
                    }),
                    new DivImage({
                        image_url: expression('@{image}'),
                        margins: {
                            left: expression('@{left}'),
                        },
                    }),
                ],
            }),
            template2: new DivGallery({
                items: [
                    new DivText({
                        text: 'string',
                        action: reference('action1'),
                    }),
                    new DivContainer({
                        items: [
                            new DivText({
                                text: reference('text2'),
                                action: { log_id: 'test', url: 'test' },
                            }),
                        ],
                    }),
                ],
            }),
            template3: new DivContainer({
                items: [
                    new DivImage({
                        height: fixed(50),
                        width: fixed(50),
                        image_url: 'imageSrc',
                    }),
                ],
                background: [
                    new DivSolidBackground({ color: '#0000000d' }),
                    new DivImageBackground({
                        image_url: reference('img'),
                        scale: reference('imgScale'),
                    }),
                    new DivLinearGradient({
                        angle: reference('bgAngle'),
                        colors: ['#99000000', '#66000000', '#33000000', '#00000000'],
                    }),
                ],
            }),
            template4: template('template3', {
                items: [
                    template('template1', {
                        vPaddings: {},
                    }),
                    template('template2'),
                ],
            }),
            template6: new DivContainer({
                items: [template('template5')],
            }),
            template5: new DivImage({
                image_url: reference('imgUrl'),
            }),
            template7: new DivContainer({
                orientation: reference('orient'),
                items: [
                    new DivContainer({
                        items: reference('title_items'),
                    }),
                ],
            }),
            template8: new DivContainer({
                items: reference('title_items'),
            }),
        };

        const helper = templateHelper(templates);

        helper.template1({
            //@ts-expect-error wrong type
            vText1: 1,
            vText2: 'test',
            vMaxlines: 2,
            vPaddings: { left: 4 },
        });

        // expected no error
        helper.template7({
            orient: 'horizontal',
            title_items: [
                new DivText({
                    text: 'hello',
                }),
            ],
        });

        helper.template7({
            orient: 'horizontal',
            //@ts-expect-error wrong type
            title_items: [1],
        });

        const block = helper.template1({
            vText1: 'test',
            vText2: 'test',
            vMaxlines: 2,
            vPaddings: { left: 4 },
        });
        expect(block).toMatchSnapshot();

        const block2 = helper.template2({
            action1: { log_id: 'action', url: 'url' },
            text2: 'text2',
        });
        expect(block2).toMatchSnapshot();

        const block3 = helper.template3({
            img: 'img',
            imgScale: 'fill',
            bgAngle: 90,
        });
        expect(block3).toMatchSnapshot();

        const block4 = helper.template4({
            img: 'img',
            imgScale: 'fill',
            vText1: '1',
            vText2: '2',
            text2: '2',
            action1: { log_id: 'action', url: 'url' },
            vMaxlines: 1,
            bgAngle: 90,
        });
        expect(block4).toMatchSnapshot();

        helper.template6({
            imgUrl: 'imgUrl',
        });

        // expression into the reference
        helper.template2({
            action1: {
                log_id: '123',
                log_url: expression('//ya.ru'),
            },
            text2: expression('1'),
        });
    });
});
