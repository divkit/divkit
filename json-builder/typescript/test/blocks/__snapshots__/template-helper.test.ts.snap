// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templateHelper should create TemplateBlock with parameters 1`] = `
{
  "type": "template1",
  "vText1": "test",
  "vText2": "test",
  "vMaxlines": 2,
  "vPaddings": {
    "left": 4
  }
}
`;

exports[`templateHelper should create TemplateBlock with parameters 2`] = `
{
  "type": "template2",
  "action1": {
    "log_id": "action",
    "url": "url"
  },
  "text2": "text2"
}
`;

exports[`templateHelper should create TemplateBlock with parameters 3`] = `
{
  "type": "template3",
  "img": "img",
  "imgScale": "fill",
  "bgAngle": 90
}
`;

exports[`templateHelper should create TemplateBlock with parameters 4`] = `
{
  "type": "template4",
  "img": "img",
  "imgScale": "fill",
  "vText1": "1",
  "vText2": "2",
  "text2": "2",
  "action1": {
    "log_id": "action",
    "url": "url"
  },
  "vMaxlines": 1,
  "bgAngle": 90
}
`;
